sw/host/opentitanlib/src/proxy/socket_server.rs

```diff
@@ -0,0 +1,298 @@
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+use super::CommandHandler;
+
+const BUFFER_SIZE: usize = 8192;
+const EOL_CODE: u8 = b'\n';
+
+fn get_next_token() -> Token {
+    static TOCKEN_COUNTER: AtomicUsize = AtomicUsize::new(0);
+    Token(TOCKEN_COUNTER.fetch_add(1, Ordering::Relaxed))
+}
+
+/// This struct listens on a TCP socket, and maintains a number of concurrent connections,
+/// receiving serialized JSON representations of `Msg`, passing them to the given
+/// `CommandHandler` to obtain responses to be sent as socket flow contol permits.  Note that
+/// this implementaion is not specific to (and does not refer to) any particular protocol.
+pub struct JsonSocketServer<Msg: DeserializeOwned + Serialize, T: CommandHandler<Msg>> {
+    command_handler: T,
+    poll: Poll,
+    socket: TcpListener,
+    socket_token: Token,
+    signals: Signals,
+    signal_token: Token,
+    connection_map: HashMap<Token, Connection>,
+    exit_requested: bool,
+    phantom: std::marker::PhantomData<Msg>,
+}
+
+impl<Msg: DeserializeOwned + Serialize, T: CommandHandler<Msg>> JsonSocketServer<Msg, T> {
+    pub fn new(command_handler: T, port: u16) -> Result<Self> {
+        let poll = Poll::new()?;
+        let socket_token = get_next_token();
+        let addr = std::net::SocketAddr::new(
+            std::net::IpAddr::V4(std::net::Ipv4Addr::new(0, 0, 0, 0)),
+            port,
+        );
+        log::info!("Setup JsonSocketServer: {}", &addr);
+        let mut socket = TcpListener::bind(addr)?;
+        poll.registry()
+            .register(&mut socket, socket_token, Interest::READABLE)?;
+        // Create a `Signals` instance that will catch given set of signals for us.
+        let signals: SignalSet = Signal::Terminate | Signal::Interrupt;
+        let mut signals = Signals::new(signals)?;
+        // And register it with our `Poll` instance.
+        let signal_token = get_next_token();
+        poll.registry()
+            .register(&mut signals, signal_token, Interest::READABLE)?;
+        Ok(Self {
+            command_handler,
+            poll,
+            socket,
+            socket_token,
+            signals,
+            signal_token,
+            connection_map: HashMap::new(),
+            exit_requested: false,
+            phantom: std::marker::PhantomData,
+        })
+    }
+
+    pub fn run_loop(&mut self) -> Result<()> {
+        let mut events = Events::with_capacity(1024);
+        while !self.exit_requested {
+            self.poll.poll(&mut events, None)?;
+            for event in events.iter() {
+                if event.token() == self.socket_token {
+                    self.process_new_connection()?;
+                } else if event.token() == self.signal_token {
+                    self.process_signals()?;
+                } else {
+                    match self.process_connection(event) {
+                        Ok(false) => (),
+                        Ok(true) => self.shutdown_connection(event)?,
```

Comment:
This can be `Ok(do_shutdown) => if do_shutdown { ... },`

Reply:
Done